rules/TypeDeclaration/Rector/ClassMethod/AddParamTypeFromPropertyTypeRector.php

```diff
@@ -0,0 +1,183 @@
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(self::ERROR_MESSAGE, [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    private string $name;
+
+    public function setName($name)
+    {
+        $this->name = $name;
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+final class SomeClass
+{
+    private string $name;
+
+    public function setName(string $name)
+    {
+        $this->name = $name;
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?ClassMethod
+    {
+        $hasChanged = false;
+
+        foreach ($node->params as $param) {
+            // already known type → skip
+            if ($param->type instanceof Node) {
+                continue;
+            }
+
+            $paramName = $this->getName($param);
+
+            // has param override? skip it
+            $hasParamOverride = (bool) $this->betterNodeFinder->findFirst(
+                $node,
+                function (Node $node) use ($paramName): bool {
+                    if (! $node instanceof Assign) {
+                        return false;
+                    }
+
+                    if (! $node->var instanceof Variable) {
+                        return false;
+                    }
+
+                    return $this->isName($node->var, $paramName);
+                }
+            );
+
+            if ($hasParamOverride) {
+                continue;
+            }
+
+            $propertyStaticTypes = $this->resolvePropertyStaticTypesByParamName($node, $paramName);
+            $possibleParamType = $this->typeFactory->createMixedPassedOrUnionType($propertyStaticTypes);
+
+            $paramType = $this->staticTypeMapper->mapPHPStanTypeToPhpParserNode($possibleParamType, TypeKind::PARAM);
+            if (! $paramType instanceof Node) {
+                continue;
+            }
+
+            $param->type = $paramType;
+            $hasChanged = true;
+        }
+
+        if ($hasChanged) {
+            return $node;
+        }
+
+        return null;
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::TYPED_PROPERTIES;
+    }
+
+    /**
+     * @return Type[]
+     */
+    private function resolvePropertyStaticTypesByParamName(ClassMethod $classMethod, string $paramName): array
+    {
+        $propertyStaticTypes = [];
+
+        $this->simpleCallableNodeTraverser->traverseNodesWithCallable($classMethod, function (Node $node) use (
+            $paramName,
+            &$propertyStaticTypes
+        ): ?int {
+            if ($node instanceof Class_ || $node instanceof Function_) {
+                // skip anonymous class and closures
```

Comment:
invalid comment, closure is not Function_, as property fetch can live inside closure

Reply:
I see. How would you phrase it?

Reply:
Skip anonymous classes and inner function